### PR TITLE
[Test Only] merge gate: add a pytest job for matmul, fused and reduction ttnn ops

### LINKED ISSUE #56004: Add pytest-based job in merge gate for testing ttnn ops
A new job is needed next to the generated kernels tests: https://github.com/tenstorrent/tt-metal/actions/runs/34397565629/job/102624616989, with some initial test coverage, reusing operations unit tests from the Sanity.

Priorities for test selection:

1. test stability (no flaky tests in merge gate)
2. short runtime
3. code coverage
4. API coverage

### PR DESCRIPTION
**Pull request by iwroszTT**
### PR Category
Test Only

### Summary
Closes #56004.

The merge gate has no pytest coverage for the matmul, fused and reduction op families. Their Sanity groups (`ttnn matmul group`, `ttnn fused group 1/2`, `ttnn reduce group`) run whole directories under 40-45 minute timeouts, so they only run nightly — a break in any of the three lands on main and is first seen when nightly sanity goes red.

This adds a `ttnn matmul fused reduction ops` entry to `tests/pipeline_reorg/ttnn_merge_gate_tests.yaml`, right next to the generated-kernels entry the issue points at, reusing those Sanity unit tests.

**No workflow file changes.** The merge gate matrix is data-driven: `merge-gate.yaml`'s `ttnn-merge-gate-tests` job calls `ttnn-smoke-tests-impl.yaml`, which reads this yaml and expands one job per (entry × SKU). `ttnn-merge-gate-tests` is already in `workflow-status`'s `needs` and `optional-jobs`.

**Selection** — 718 tests, by test function so it stays small as those files grow:

| family | coverage |
| --- | --- |
| matmul | matched width/height (2D and 4D), dot product, transpose a/b, tutorial shapes, optional output tensor, small-shape pcc, sharded (2D mcast, mcast in0/in1, bmm), linear with a core grid / fp32 dest acc + bias / bloom ff2 |
| fused | softmax over every dim of a 3D input, layer_norm with and without weight and bias, rms_norm, group_norm height-sharded and with optional weight/bias, batch_norm queue-id and preallocated-output overloads |
| reduction | sum, mean, max over each dim of a 2D input, global min, argmax tie order, argmax into a preallocated output |
| program cache | the matmul and reduce cache-key tests, which catch a wrong hash silently reusing a program built for another shape or dtype |

**Measured locally on a P150:** 718 passed on five consecutive runs; 57s of pytest / 66s of wall-clock with a warm kernel cache, 83s / 92s with a partly cold one. Slowest single test is 1.2s warm, inside the merge gate's 5s-per-test-case rule. Timeout set to 10 min per SKU, the same budget shape as the existing `ttnn eltwise infra` entry (753 tests, 115s in CI).

**Left out**, for the issue's stability-then-runtime priorities:

- `test_topk` — the full matrix is ~4 minutes, and its tie order is sensitive to index packing
- `cumsum` and the full `test_argmax` matrix — ~1.5 minutes each
- the large-shape matmul cases — `test_matmul_with_core_grid[batch_size=7]` alone takes 10s on a cold kernel cache, breaking the 5s-per-test rule

### Notes for reviewers

- **SKUs.** The issue asked for P150 and N150; the entry uses `bh_p150b_civ2_viommu` and `wh_n150_civ2`, the CIv2 merge-gate runners for those parts, matching `llk_merge_gate_tests.yaml` and `triage_merge_gate_tests.yaml`. The bare `bh_p150`/`wh_n150` SKUs are bare-metal sanity runners with no `ttnn.merge_gate` time budget, so they would fail the budget check. On `merge_group` the blackhole leg rewrites to `bh_p150b_civ2_viommu_prio`. Both legs fit the existing 45-min budgets, so `time_budget.yaml` is unchanged.
- **Owner.** `owner_id` is set to Borys Bradel, who owns all three source Sanity groups. Say so if the pings should go elsewhere.
- **Hardware only**, no `sim_*` SKUs, to keep the gate cheap — same reasoning as the codegen-ops entry above it.
- **One pytest invocation on purpose.** Device bring-up is ~7s, so splitting the three families into separate entries would cost more than the tests do.
- Local validation: `verify_time_budget.py` and `prepare_test_matrix.py` both clean for `pull_request` and `merge_group`, and all pre-commit hooks pass. The first CI run on this PR is the real check on the two SKUs, since the timings above are from a Blackhole dev box.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:iwrosz/56004-ttnn-ops-merge-gate)

### CODE REVIEW

**Comment by blozano-tt** [comment]
I love this initiative! A few suggestions to consider.

1. Tests could be selected from a separate yaml, or markers could be used to keep the workflow simple. This would be helpful from a codeowner perspective. Devs can own the test list file. 

2. Aditi Shah has a new dashboard that highlights ND tests to avoid 
3. I’d love to see tests that are added to merge gate be disabled in sanity tests or shifted to a simulation sku. 

I had a dream that all non flaky ttnn unit tests would get shifted to merge gate, and sanity tests become majority simulation only.

**Comment by iwroszTT** [comment]
> I love this initiative! A few suggestions to consider.
> 
> 1. Tests could be selected from a separate yaml, or markers could be used to keep the workflow simple. This would be helpful from a codeowner perspective. Devs can own the test list file.
> 2. Aditi Shah has a new dashboard that highlights ND tests to avoid
> 3. I’d love to see tests that are added to merge gate be disabled in sanity tests or shifted to a simulation sku.
> 
> I had a dream that all non flaky ttnn unit tests would get shifted to merge gate, and sanity tests become majority simulation only.

Hi @blozano-tt, thanks for the suggestions. I've changed the test selection to a marker's based.

File: tests/pipeline_reorg/ttnn_merge_gate_tests.yaml [hunk 1]
```diff
@@ -139,3 +139,22 @@
...
   budget_type: merge_gate
   owner_id: U09BX3N0N95 # Mateusz Nowak
+
+# Matmul, fused and reduction coverage, reusing the Sanity op groups.
+#
+# The selection lives in the test files, as @pytest.mark.merge_gate (registered in pytest.ini),
+# not as a list of node IDs here.
+- name: "ttnn matmul fused reduction ops"
+  cmd: >-
+    pytest --timeout 60 -xv -m merge_gate
```
[the remark this answers is not in the document]
Comment (iwroszTT) [reply tests/pipeline_reorg/ttnn_merge_gate_tests.yaml:149]: It's assumed there will be no flaky tests -- any unstable test cases will be removed.

[the remark this answers is not in the document]
Comment (iwroszTT) [reply tests/pipeline_reorg/ttnn_merge_gate_tests.yaml:149]: If there's a problem with the PR it should fail fast and unblock the merge queue.

**Review by roseli-TT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]
